src/librustc_trait_selection/traits/error_reporting/mod.rs

```diff
@@ -1840,55 +1824,44 @@ pub fn suggest_constraining_type_param(
             }
         }
 
-        let where_clause_span =
-            generics.where_clause.span_for_predicates_or_empty_place().shrink_to_hi();
+        let where_clause_span = generics.where_clause.span_for_predicates_or_empty_place();
+        // Account for `fn foo<T>(t: T) where T: Foo,` so we don't suggest two trailing commas.
+        let mut trailing_comma = false;
+        if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(where_clause_span) {
+            if snippet.ends_with(",") {
+                trailing_comma = true;
+            }
+        }
+        let where_clause_span = if trailing_comma {
+            let hi = where_clause_span.hi();
+            Span::new(hi - BytePos(1), hi, where_clause_span.ctxt())
```

Comment:
It would be good if we could write this as `span.offset_sub(1)` instead. Also, have you considered utf8 and stuff wrt. ICEs?

Reply:
> Also, have you considered utf8 and stuff wrt. ICEs?

The only thing that could push that span is a `,` char, so this should never fail. (Famous last words.)